Paging test moved from wormtable to simulated backend.

macieksmuga opened the pull request:
New `TestAbstractBackend` class introduced in `unit/test_backends.py`,
gathers together backend implementation independent logic,
`TestSimulatedBackend` and `TestWormtableBackend` classes inherit from it.
`TestPaging` is a new subclass of `TestSimulatedBackend`. This setup
should provide a clean migration path out of Wormtable for testing backend.

 For now, I could only pull `TestVariantSets.testPagination()` out of 
`TestWormtableBackend`, as `SimulatedBackend` only provides mock objects
for `VariantSet`s. As `SimulatedBackend` grows, the other tests in `test_backends.py`
can be pulled out as well.

I could technically pull out `TestVariantSets.testIds()` now as well, but this doesn't fit
the new `TestPaging` class's focus on pagination. 

Addresses issue #241

jeromekelleher commented on tests/unit/test_backends.py:
```diff
@@ -93,26 +93,10 @@ def convertInfoValue(column, value):
-    def tearDown(self):
-        for table in self._tables.values():
-            table.close()
+        self._backend = backend.AbstractBackend()
```
I don't think there's much point in instantiating an AbstractBackend here, as this will get over-written in the subclass anyway. We can just get rid of `setUp` here.

jeromekelleher commented on tests/unit/test_backends.py:
```diff
@@ -177,6 +161,53 @@ def getCallSets(self, variantSetId, pageSize=100):
+            results.append(variantSetIds)
+        for result in results[1:]:
+            self.assertEqual(result, results[0])
+        # TODO repeat this for full object equality, not just ids
```
We might as well just do this for object equality now and get rid of this comment. We can replace this with something like
```python
results = []
for pageSize in range(1, 10):
    results.append(list(self.getVariantSets(pageSize=pageSize)))
for result in results[1:]:
    self.assertEqual(result, results[0])
```

jeromekelleher commented on tests/unit/test_backends.py:
```diff
@@ -177,6 +161,53 @@ def getCallSets(self, variantSetId, pageSize=100):
+            numCalls=100, numVariantSets=10)
+        # TODO arbitrary values
+
+    def tearDown(self):
```
Might as well delete this if it doesn't do anything.

jeromekelleher commented on tests/unit/test_backends.py:
```diff
@@ -177,6 +161,53 @@ def getCallSets(self, variantSetId, pageSize=100):
+        pass
+
+
+class TestPaging(TestSimulatedBackend):
```
Some docstring comments would be nice. Just a quick sentence explaining what we want to achieve in this test.

jeromekelleher commented:
This looks good, thanks @macieksmuga. I wonder if there's much point in trying to make a general superclass of our new tests and the Wormtable backend test --- we're aiming to delete TestWormtableBackend, so there's not much point in trying to share code nicely with it. 

The approach I would suggest here is to make a new class, `TestBackend(unittest.TestCase)` that is completely separate from the current inheritence tree. This should create a simulated backend, and test the code in backend.py. This means tapping in to the top-level calls like `searchVariantSets` and verifying that they return the same values as we would get if we accessed the VariantSets directly. That is, we run `self._backend.searchVariantSets()`, and compare this with `self._backend.getVariantSets()`.

Paging is actually the main thing we want to test within `backend.py`, so I think we should be a bit more comprehensive: the number of elements returned should always be <= the requested page size, we should never return 0 length pages, as well as the consistency checks we're doing here.

I guess this has taken us a long way from your actual PR - sorry! What you have done here is good and we can merge this, and move on with adding more tests if you prefer. I just thought it would be good to explain what I think we're trying to test in this module now (in the past, it was basically everything).

macieksmuga commented:
@jeromekelleher There were a couple of possible directions I could go in to resolve the issue, and this PR is so far a quick sketch to get your take on the matter. Thus I really appreciate your comments above.f

As you mentioned, up to this point this module has tested... kind of everything (including data-driven-like tests). Thus, I am sympathetic to your suggestion to convert this to a collection of paging-oriented tests against `SimulatedBackend`, and unless you comment further on the below issue/suggestion, I will suspend this PR for now and implement what you suggested above.

But I have one conceptual problem with this approach, and it's the reason why I structured the test class hierarchy the way I did: such a testing regimen would then exclude unit-test coverage of the `FileSystemBackend` class. Admittedly this is a pretty lightweight class that mostly hooks up lower-level access to variants, reads and references (with the bulk of the code now in `AbstractBackend`), but I still felt uncomfortable cutting it out of the unit tests.

With the current structure in `test_backends.py`, I could write unit tests specifically against methods in every class in the `backend` hierarchy separately. The next step of this approach would be to transition `TestWormtableBackend` to `TestFileSystemBackend`, and switch its setUp method to generate/point at the `.tbi` indices, an approach I believe @dcolligan started to investigate.

jeromekelleher commented:
Ah, now I get it! Yes, everything you're saying here makes sense @macieksmuga, and I agree it would be nice to get some sort of coverage on the `FileSystemBackend` class. We can make a dataset out of the test data in `tests/data` which has exactly the right structure and all the files are already indexed. I take back the points about the class hierarchy above, you're spot on.

macieksmuga commented:
Thanks @jeromekelleher! I'll go ahead with the current structure, beef up `SimulatedBackend` to enable more tests against it (that effectively end up testing `AbstractBackend`, I belatedly realized), and fill out the gaps you've pointed out elsewhere.

macieksmuga commented:
Thinking about this a little bit more, I'll just make `TestAbstractBackend` use the `SimulatedBackend` and remove the now redundant `TestSimulatedBackend` class altogether (after all, the purpose of `SimulatedBacked` was precisely to enable running unit tests against the methods in `AbstractBackend`)

jeromekelleher commented:
Sounds great @macieksmuga, I'm looking forward to seeing the results!

macieksmuga commented:
@jeromekelleher I spent some time trying to implement what I outlined above, and fairly quickly discovered that many of the existing tests in `test_backends.py` are indeed more data-driven than unit: For example, switching over to a `pysam.VariantFile` backed `TestFileSystemBackend` ran right into the current issue with `HtslibVariantSet.getVariants()` not implementing restriction by `callSetIds`.

With this experience behind me, I'm now inclined to follow your initial proposal, namely unit-test all the methods implemented in `AbstractBackend` with an emphasis on paging, and leave coverage in the unit-tests of the `FileSystemBackend` to a simple check that it loads OK, and that the various `_*SetIds` are correct.

macieksmuga commented:
@jeromekelleher I've implemented all the unit tests I can for `backends.py`, but am running across this:
```
AttributeError: 'HtslibVariantSet' object has no attribute 'getCallSets'
```
This is a method that's instantiated both in the `Simulated-` and `WormtableVariantSet`s, so should have at least a trivial implementation in `HtslibVariantSet` (that possibly throws a `NotImplementedException` ? - I added it in for now).

Anyway, I'd strongly prefer to leave this test in and have somebody (perhaps myself) write the actual missing method rather than try to excise the test out, even though it will break the CI build. Let me know what you think about this.

jeromekelleher commented on tests/unit/test_backends.py:
```diff
@@ -22,6 +23,211 @@
+            protocol.GASearchCallSetsResponse, "callSets")
+
+    def testGetVariantSets(self):
+        """
```
Minor nit: comments for test cases shouldn't be docstrings. See STYLE.md

jeromekelleher commented on tests/unit/test_backends.py:
```diff
@@ -22,6 +23,211 @@
+            self._backend.variantSetsGenerator)
+        response = protocol.GASearchVariantSetsResponse.fromJsonString(
+            responseStr)
+        self.assertTrue(type(response) is protocol.GASearchVariantSetsResponse)
```
Use `isinstance` here.

jeromekelleher commented on tests/unit/test_backends.py:
```diff
@@ -22,6 +23,211 @@
+        responseStr = self._backend.searchVariantSets(request.toJsonString())
+        response = protocol.GASearchVariantSetsResponse.fromJsonString(
+            responseStr)
+        self.assertTrue(type(response) is protocol.GASearchVariantSetsResponse)
```
ditto here.

jeromekelleher commented on tests/unit/test_backends.py:
```diff
@@ -22,6 +23,211 @@
+                    pageSize=pageSize)]
+            results.append(variantSetIds)
+        for result in results[1:]:
+            print(result)
```
remove this print please

jeromekelleher commented:
This looks great, thanks @macieksmuga. For `getCallSets` I would add a trivial implementation in the superclass so your tests pass. This is being tackled in #286.

macieksmuga commented:
thanks @jeromekelleher! pushed new version with your corrections.

jeromekelleher commented:
Great! +1, assuming CI passes.

macieksmuga commented:
yeah, well, 'CORS it won't ;)
